cmd/tls-checker/main.go

```diff
@@ -0,0 +1,430 @@
+	flag.String("key", "", "path to file of base64-encoded private key for client TLS")
+	flag.String("key-file", "", "path to file of base64-encoded private key for client TLS")
+	flag.String("cert", "", "base64-encoded public key for client TLS")
+	flag.String("cert-file", "", "path to file of base64-encoded public key for client TLS")
+	flag.String("ca", "", "base64-encoded certificate authority for mutual TLS")
+	flag.String("ca-file", "", "path to file of base64-encoded certificate authority for mutual TLS")
+	flag.String("status-codes", "200", "valid response status codes")
+	flag.Bool("skip-verify", false, "skip certifiate validation")
+	flag.Duration("timeout", 5*time.Minute, "timeout duration")
+	flag.Bool("exit-on-error", false, "exit on first tls check error")
+	flag.String("log-env", "development", "logging config: development or production")
+	flag.String("log-level", "error", "log level: debug, info, warn, error, dpanic, panic, or fatal")
+	flag.Bool("verbose", false, "output extra information")
+
+	flag.Parse(os.Args[1:])
+
+	v := viper.New()
+	v.BindPFlags(flag)
+	v.SetEnvPrefix("HEALTHCHECKER")
+	v.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+	v.AutomaticEnv()
+
+	tlsCheckErrors := make([]error, 0)
+	defer func() {
+		if len(tlsCheckErrors) > 0 {
+			os.Exit(1)
+		}
+	}()
+
+	logger, err := createLogger(v.GetString("log-env"), v.GetString("log-level"))
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+
+	defer logger.Sync()
+
+	err = checkConfig(v)
+	if err != nil {
+		switch e := err.(type) {
+		case *errInvalidPath:
+			logger.Fatal(e.Error(), zap.String("path", e.Path))
+		case *errInvalidScheme:
+			logger.Fatal(e.Error(), zap.String("scheme", e.Scheme))
+		case *errInvalidStatusCode:
+			logger.Fatal(e.Error(), zap.Int("code", e.Code))
+		}
+		logger.Fatal(err.Error())
+	}
+
+	verbose := v.GetBool("verbose")
+	schemes := strings.Split(strings.TrimSpace(v.GetString("schemes")), ",")
+	hosts := strings.Split(strings.TrimSpace(v.GetString("hosts")), ",")
+	paths := strings.Split(strings.TrimSpace(v.GetString("paths")), ",")
+	statusCodes, _ := stringSliceToIntSlice(strings.Split(strings.TrimSpace(v.GetString("status-codes")), ","))
+
+	// TLS Versions that should not work
+	var invalidTLSVersions = []uint16{
+		tls.VersionTLS10,
+		tls.VersionTLS11,
+	}
+
+	for _, tlsVersion := range invalidTLSVersions {
+
+		tlsName := getTLSName(tlsVersion)
+
+		httpClient, err := createHTTPClient(v, logger, tlsVersion)
+		if err != nil {
+			logger.Fatal(errors.Wrap(err, "error creating http client").Error())
+		}
+
+		exitOnError := v.GetBool("exit-on-error")
+
+		for _, scheme := range schemes {
+			for _, host := range hosts {
+				for _, path := range paths {
+					url := scheme + "://" + host + path
+					if verbose {
+						logger.Info("checking url", zap.String("url", url), zap.String("tlsVersion", tlsName))
+					}
+					err := checkURL(httpClient, url, statusCodes, logger)
```

Comment:
since we are using invalid `tlsversion` for creating http server wouldn't this always fail? @chrisgilmerproj Think I am missing something.

Reply:
In that method I pass back an error if it succeeds and nil if it doesn't succeed. I could change the function name if that helps:
```suggestion
					err := checkURLWontConnect(httpClient, url, statusCodes, logger)
```